# Meeting Agenda
1. Go over the completed team proposal sections and rationales
2. Discuss if diet could also be worth investigating as many MS patients try to go on a "less inflammation-inducing" diet (low in saturated fats) and might tie into metabolites found after functional analysis (could be aim4)
3. Address denoising/clustering step issue


# Meeting Notes
1. Overview of proposal so far:
   a) General information about MS, looking into different types of MS and their symptoms, and how this links to the gut microbiome.
   b) The question was streamlined since our last meeting.
   c) The hypothesis might focus on the functional or profile differences between the groups. It sounds good as is though.
     i) The null hypothesis could be that there are no differences between the three groups.
     ii) We might want to include a predictive element (predicting disease progression based on gut microbiome). It is not essential but it depends on how we want to frame our project. This can also change after the proposal deadline. We don't necessarily need to anlayse the predictive aspect, it can be one paragraph at the end of the project.
     iii) One option is to subset our dataset but this might result in even fewer samples than we already have.
  d) For the first aim, we may also separate our samples by untreated and treated.
  e) We can include a fourth aim, if we see any differences between disease progression - if these findings are applicable to other datasets.
  f) We can consider weighted and unweighted UNIFRAC for analysis. Unweighted is quick to do and can give some good information.
2. It is important to explain that the reason we want to investigate the difference between PPMS and SPMS is because the authors of the original study combined them together but did not really explain why. However, they are significantly different symptom-wise.
3. Keep in mind the limitations of a functional analysis: it is a prediction, so everything discovered performed with picrust2 is only the potential of microbes, not what is specifically happening in the microbiome. We do not describe the actual function of a community. Our resolution (16S) is not very high, meaning that much of our taxonomic information stops at the genus level but within a genus species can be very different (even if we get to the species level bacteria can vary greatly).
   a) Be careful in this section of using "strong" sentences.
