##### Install packages #####
# Start by installing all necessary packages when asked if you want to install
# from source, please just type Yes in the terminal below

# If you don't have BiocManager, here is the code to install it
# A lot of you probably already have this so you can skip
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

# Create a list of all the packages you need to install
pkgs <- c("ALDEx2", "SummarizedExperiment", "Biobase", "devtools", 
          "ComplexHeatmap", "BiocGenerics", "metagenomeSeq", 
          "Maaslin2", "edgeR", "lefser", "limma", "KEGGREST", "DESeq2")

# Use the above list to install all the packages using a for loop
for (pkg in pkgs) {
  if (!requireNamespace(pkg, quietly = TRUE))
    BiocManager::install(pkg)
}
# when asked if you want to update all, some or none please type "n" for none

# After installing all of its above dependencies, install ggpicrust2
install.packages("ggpicrust2")

#### Load packages ####
# Load all necessary libraries
library(readr)
library(ggpicrust2)
library(tibble)
library(tidyverse)
library(ggprism)
library(patchwork)
library(DESeq2)
library(ggh4x)
library(dplyr)


#### Import files and preparing tables ####
#Importing the pathway PICrust2
abundance_file <- "pathway_abundance.tsv"
abundance_data <- read_delim(abundance_file, delim = "\t", col_names = TRUE, trim_ws = TRUE, skip = 1)
abundance_data  =as.data.frame(abundance_data)
colnames(abundance_data)[1] <- "pathways"

#Import your metadata file, no need to filter yet
metadata <- read_delim("corrected_ms_metadata.tsv")

#Example Looking at subject number
#If you have multiple variants, filter your metadata to include only 2 at a time

#Remove NAs for your column of interest in this case subject
metadata = metadata[!is.na(metadata$disease_course),]

SPMS_vs_control <- c("SPMS", "Control")
PPMS_vs_control <- c("PPMS", "Control")
RRMS_vs_control <- c("RRMS", "Control")
SPMS_vs_PPMS <- c("SPMS", "PPMS")
SPMS_vs_RRMS <- c("SPMS", "RRMS")
PPMS_vs_RRMS <- C("PPMS", "RRMS")

metadata_svc <- metadata[metadata$disease_course %in% SPMS_vs_control, ]

#Starting with SPMS vs control
#Filtering the abundance table to only include samples that are in the filtered metadata
sample_names_svc = metadata_svc$'sample-id'
sample_names_svc = append(sample_names_svc, "pathways")
abundance_data_filtered_svc = abundance_data[, colnames(abundance_data) %in% sample_names_svc] #This step is the actual filtering

#Removing individuals with no data that caused a problem for pathways_daa()
abundance_data_filtered_svc =  abundance_data_filtered_svc[, colSums(abundance_data_filtered_svc != 0) > 0]

#Ensuring the rownames for the abundance_data_filtered is empty. This is required for their functions to run.
rownames(abundance_data_filtered_svc) = NULL

#verify samples in metadata match samples in abundance_data
abun_samples_svc = rownames(t(abundance_data_filtered_svc[,-1])) #Getting a list of the sample names in the newly filtered abundance data
metadata_svc = metadata_svc[metadata_svc$`sample-id` %in% abun_samples_svc,] #making sure the filtered metadata only includes these samples

#### DESEq ####
#Perform pathway DAA using DESEQ2 method
abundance_daa_results_df_svc <- pathway_daa(abundance = abundance_data_filtered_svc %>% column_to_rownames("pathways"), 
                                        metadata = metadata, group = "disease_course", daa_method = "DESeq2")

# Annotate MetaCyc pathway so they are more descriptive
metacyc_daa_annotated_results_df_svc <- pathway_annotation(pathway = "MetaCyc", 
                                                       daa_results_df = abundance_daa_results_df_svc, ko_to_kegg = FALSE)

# Filter p-values to only significant ones
feature_with_p_0.05_svc <- metacyc_daa_annotated_results_df_svc %>% filter(p_values < 0.05)

#Changing the pathway column to description for the results 
feature_desc_svc = inner_join(feature_with_p_0.05_svc,metacyc_daa_annotated_results_df_svc, by = "feature", copy = FALSE)
feature_desc_svc$feature = feature_desc_svc$description
feature_desc_svc = feature_desc_svc[,c(1:7)]
colnames(feature_desc_svc) = colnames(feature_with_p_0.05_svc)

#Changing the pathway column to description for the abundance table
abundance_svc = abundance_data_filtered_svc %>% filter(pathways %in% feature_with_p_0.05_svc$feature)
colnames(abundance_svc)[1] = "feature"
abundance_desc_svc = inner_join(abundance_svc,metacyc_daa_annotated_results_df_svc, by = "feature")
abundance_desc_svc$feature = abundance_desc_svc$description
#this line will change for each dataset. 34 represents the number of samples in the filtered abundance table
abundance_desc_svc = abundance_desc_svc[,-c(520:ncol(abundance_desc_svc))] 

# Generate a heatmap
pathway_heatmap(abundance = abundance_desc_svc %>% column_to_rownames("feature"), metadata = metadata_svc, group = "disease_course")

# Generate pathway PCA plot
pathway_pca(abundance = abundance_data_filtered %>% column_to_rownames("pathway"), metadata = metadata, group = "subject")

# Generating a bar plot representing log2FC from the custom deseq2 function

# Go to the Deseq2 function script and update the metadata category of interest

# Lead the function in
source("DESeq2_function.R")

# Run the function on your own data
res =  DEseq2_function(abundance_data_filtered, metadata, "subject")
res$feature =rownames(res)
res_desc = inner_join(res,metacyc_daa_annotated_results_df, by = "feature")
res_desc = res_desc[, -c(8:13)]
View(res_desc)

# Filter to only include significant pathways
sig_res = res_desc %>%
  filter(pvalue < 0.05)
# You can also filter by Log2fold change

sig_res <- sig_res[order(sig_res$log2FoldChange),]
ggplot(data = sig_res, aes(y = reorder(description, sort(as.numeric(log2FoldChange))), x= log2FoldChange, fill = pvalue))+
  geom_bar(stat = "identity")+ 
  theme_bw()+
  labs(x = "Log2FoldChange", y="Pathways")
